Global styles revisions: add a reset to default revision

ramonjd opened the pull request:
Resolves https://github.com/WordPress/gutenberg/issues/52951

## What?
Inserting a reset object as the first "revision" to reset to theme defaults.

Originally spoken about in https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1532324313

## Why?
Testing out reducing any perceived confusion explained in https://github.com/WordPress/gutenberg/issues/52951

Basically we are trying to find a way to show the revisions panel after a the first change, instead of two changes.

Reason behind the two change rule here: https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1532276020

## How?
Adding a 'fake' revision to the end of the revisions UI list.

## TODO

- [x] Fix tests, add an E2E test

## Testing Instructions

1.  Using a block theme, head to the site editor
2. Create a single global styles change and save
3. Open the revisions panel
4. You should be see 1 revision and the original theme "revision"
5. Check that selecting revisions works as per usual
6. Change themes to a new block theme and add a few revisions. Make sure the theme name is correctly rendered in the reset styles revision button at the bottom of the styles revisions list.

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/6458278/dea3f22b-0b90-4102-869e-98255c1411c7

ramonjd committed:
Inserting a reset object as the first "revision" to reset to theme defaults

ramonjd committed:
Update tests

ramonjd commented:
I'm wonder if this will make the "Reset to default" button in the drop down redundant?

<img width="378" alt="Screenshot 2023-07-26 at 4 59 37 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/8131f683-0a05-40d1-ade3-100bd26e8367">

Maybe if these changes went through, it might be okay to leave both in there for a while for easy access? 🤷🏻

andrewserong commented:
Love the idea! I can take this for a proper spin tomorrow 🙂

> Maybe if these changes went through, it might be okay to leave both in there for a while for easy access? 🤷🏻

Being able to access from two places sounds good to me, especially since it's nice to be able to reset things quickly if you need to without having to go digging?

ramonjd committed:
Update tests

ramonjd commented:
There's a bug here somewhere:

- Add a few revisions
- Reset to defaults and save
- Create a new revision

For some reason the reset revision is lost. 

_Edit_: fixed in [f3e39e2](https://github.com/WordPress/gutenberg/pull/52965/commits/f3e39e2e864d991dfd79a20e7761dce676b5cc67)

jameskoster commented:
Thanks for opening the PR, this is a good idea imo. 

I appreciate it's not technically a 'revision' but I suspect users will interpret it as one, so we can keep special treatment to a minimum. For example the button label can remain "Apply" instead of "Reset to defaults".

The timeline item label can just be "Default styles" – I think this would harmonise with the "Reset to defaults" action in the Styles panel header.

ramonjd committed:
- Use rendered Theme name as the revisions button
- Ensure that any undefined properties, e.g., styles, settings, are set as empty objects so that any theme defaults are correctly displayed by default

ramonjd committed:
Added a test to verify reset UI

andrewserong reviewed (approved):
Nice work, this is testing well for me:

✅ Displays correctly when there's at least 1 revision
✅ For themes with heaps and heaps of revisions, it displays at the very bottom of the list
✅ Selecting revisions works as usual
✅ Resetting to the default styles works as expected

https://github.com/WordPress/gutenberg/assets/14988353/d36d1bdf-fb33-4e07-be0e-a8a75a1a9ca5

This LGTM and feels like a nice addition to the UI! ✨

ramonjd commented:
Thanks for testing @andrewserong!
